optimize implementation of Zip::fold and company

sarah-quinones opened the pull request:
adds a more efficient implementation of `core::iter::Zip::{fold, rfold, try_fold, try_rfold}`  
on my machine, this gives a 15% speedup on the `iter::bench_skip_cycle_skip_zip_add_sum` benchmark, and a 32% speedup on `iter::bench_skip_then_zip`.

i haven't noticed any performance regressions

Commit:
optimize implementation of Zip::fold and company

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @m-ou-se (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

timvermeulen commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -171,26 +246,109 @@ macro_rules! zip_impl_general_defaults {
             }
         }
+
+        #[inline]
+        default fn fold<T, F>(self, init: T, mut f: F) -> T
+        where
+            F: FnMut(T, Self::Item) -> T,
+        {
+            let mut a = self.a;
+            let mut b = self.b;
+
+            let acc = a.try_fold(init, move |acc, x| match b.next() {
+                Some(y) => Ok(f(acc, (x, y))),
+                None => Err(acc),
+            });
+
+            match acc {
+                Ok(exhausted_a) => exhausted_a,
+                Err(exhausted_b) => exhausted_b,
+            }
+        }
```
When a `fold` implementation can't be implemented on top of an inner `fold`, it's common in libcore to delegate to `self`'s `try_fold` instead in order to avoid some duplicate logic. See e.g. [`Take::fold`](https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/take.rs#L101-L113). `TakeWhile`, `MapWhile`, and `Scan` do the same thing. You could see if it affects benchmark results in any way, but I'd hope that it won't.

SIde note: closures in the implementation of iterator adapters are typically not written outright, but returned from a nested function, see #62429.

timvermeulen commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -171,26 +246,109 @@ macro_rules! zip_impl_general_defaults {
+            let a = &mut self.a;
+            let b = &mut self.b;
+
+            let acc = a.try_fold(init, move |acc, x| match b.next() {
+                Some(y) => {
+                    let result = f(acc, (x, y));
+                    match result.branch() {
+                        ControlFlow::Continue(continue_) => Ok(continue_),
+                        ControlFlow::Break(break_) => Err(R::from_residual(break_)),
+                    }
+                }
+                None => Err(R::from_output(acc)),
+            });
```
This pattern of either continuing with an output value or breaking with an output value or a residual value can be expressed a lot simpler by the `ControlFlow` enum, and specifically its (non-public) `from_try` and `into_try` methods. [`Take::try_fold`](https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/take.rs#L75-L99) illustrates this really well.

timvermeulen commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -171,26 +246,109 @@ macro_rules! zip_impl_general_defaults {
+            }
+        }
+
+        #[inline]
+        default fn rfold<T, F>(mut self, init: T, mut f: F) -> T
+        where
+            A: DoubleEndedIterator + ExactSizeIterator,
+            B: DoubleEndedIterator + ExactSizeIterator,
+            F: FnMut(T, Self::Item) -> T,
+        {
+            self.adjust_back();
+            let mut a = self.a;
+            let mut b = self.b;
+
+            let acc = a.try_rfold(init, move |acc, x| match b.next_back() {
+                Some(y) => Ok(f(acc, (x, y))),
+                None => Err(acc),
+            });
+
+            match acc {
+                Ok(exhausted_a) => exhausted_a,
+                Err(exhausted_b) => exhausted_b,
+            }
+        }
```
Here we know that the inner iterators have the same length (after adjusting), so it might be worth writing this in terms of `self.a.rfold` instead. We'd have to panic in case `b.next_back` returns `None`. I have not benchmarked this, but in theory `rfold` might lead to more efficient machine code, depending on the underlying iterator.

sarah-ek replied:
this gave me the idea of applying a similar optimization to `fold`. by checking if both iterators have the same length (which i assumed is a common case for `zip`), we can use `a.fold` instead of `a.try_fold`

timvermeulen replied:
Very cool! I didn't think about doing that with `size_hint`.

timvermeulen reviewed (commented):
This is an excellent first contribution 👍

I've left some initial feedback. `Zip` is weird in the sense that it needs to choose which inner iterator to iterate internally and which one externally, but to me choosing the first one to iterate internally seems like a very reasonable choice.

Commit:
apply review suggestions, optimize trusted random access impl

sarah-ek commented:
thanks for the feedback! i've applied your suggestions to the code, and additionally made some further optimizations. (though they don't show a difference on the current zip benchmarks, which are somewhat limited)

the8472 reviewed (commented):
Are there tests that target the new unsafe `TrustedRandomAccess` implementations? If not we need them because that particular specialization has a long history of unsoundness.

sarah-ek commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -171,26 +309,58 @@ macro_rules! zip_impl_general_defaults {
+            F: FnMut(T, Self::Item) -> T,
+        {
+            let (lower, upper) = Iterator::size_hint(&self);
+            if upper == Some(lower) {
+                self.a.fold(init, check_fold(self.b, f))
+            } else {
+                ZipImpl::try_fold(&mut self, init, ok(f)).unwrap()
```
fair point about the binary bloat. i'm okay with undoing this change since i don't think going from `fold` to `try_fold` is going to be *that* bad

sarah-ek replied:
would it be a good idea to add tests that would fail with a faulty `TrustedRandomAccess` impl? what kind of tests should i be adding?

sarah-ek replied:
i removed the double implementation thingy, and added some tests for the new unsafe implementations.

Commit:
apply review suggestions:

 - remove size_hint check in fold
 - add tests for new unsafe fold implementations

the8472 commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -253,6 +407,40 @@ where
+    if A::MAY_HAVE_SIDE_EFFECT || B::MAY_HAVE_SIDE_EFFECT {
+        let sz_a = zipped.a.size();
+        let sz_b = zipped.b.size();
+        // Adjust a, b to equal length, make sure that only the first call
+        // of `next_back` does this, otherwise we will break the restriction
+        // on calls to `zipped.next_back()` after calling `get_unchecked()`.
```
Since this has been extracted to a function this comment can be made a doc comment for that function.

scottmcm commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -171,26 +298,53 @@ macro_rules! zip_impl_general_defaults {
+        where
+            F: FnMut(T, Self::Item) -> T,
+        {
+            ZipImpl::try_fold(&mut self, init, ok(f)).unwrap()
```
Yup, as [the8472](https://github.com/the8472) said,
```suggestion
            ZipImpl::try_fold(&mut self, init, NeverShortCircuit::wrap_mut_2(f)).0
```
so that neither the compiler nor the reader needs to spend time proving that it's not `Err`.

scottmcm commented on library/core/src/iter/adapters/zip.rs:
```diff
@@ -129,12 +185,83 @@ trait ZipImpl<A, B> {
+}
+
+#[inline]
+fn check_try_fold<
```
YMMV: this is a very long generic arguments list, so consider applying some APIT.  For example, maybe it could be

```rust
fn check_try_fold<
    'b,
    AItem,
    BItem,
    T,
    R: Try<Output = T>,
>(
    b: &'b mut impl Iterator<Item = BItem>,
    mut f: impl 'b + FnMut(T, (AItem, B::Item)) -> R,
) -> impl 'b + FnMut(T, AItem) -> ControlFlow<R, T> {
```

(Up to whether you think it's better, though.)

Commit:
apply review suggestions

sarah-ek commented:
i've applied the previous suggestions. i tried to rebase on the master branch to be able to get relevant comparisons from benchmarks. i hope i didn't mess up any git stuff

saethlin commented:
I'm curious
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3791353368/jobs/6446718408) [(plain)](https://github.com/rust-lang-ci/rust/commit/b2afce9221be4361000049c6fe8b94397ba80354/checks/10328103392/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling profiler_builtins v0.0.0 (/checkout/library/profiler_builtins)
[RUSTC-TIMING] build_script_build test:false 0.159
[RUSTC-TIMING] build_script_build test:false 0.199
[RUSTC-TIMING] build_script_build test:false 0.432
error[E0599]: no function or associated item named `wrap_mut_2` found for struct `NeverShortCircuit` in the current scope
    |
    |
251 | / macro_rules! zip_impl_general_defaults {
252 | |     () => {
253 | |         default fn new(a: A, b: B) -> Self {
254 | |             Zip {
...   |
298 | |             ZipImpl::try_fold(&mut self, init, NeverShortCircuit::wrap_mut_2(f)).0
    | |                                                                   |
    | |                                                                   |
    | |                                                                   function or associated item not found in `NeverShortCircuit<_>`
    | |                                                                   help: there is an associated function with a similar name: `wrap_mut_2_imp`
332 | |     };
333 | | }
333 | | }
    | |_- in this expansion of `zip_impl_general_defaults!`
...
344 |       zip_impl_general_defaults! {}
    |
   ::: library/core/src/ops/try_trait.rs:379:1
    |
    |
379 |   pub(crate) struct NeverShortCircuit<T>(pub T);
    |   -------------------------------------- function or associated item `wrap_mut_2` not found for this struct

error[E0599]: no function or associated item named `wrap_mut_2` found for struct `NeverShortCircuit` in the current scope
    |
    |
251 | / macro_rules! zip_impl_general_defaults {
252 | |     () => {
253 | |         default fn new(a: A, b: B) -> Self {
254 | |             Zip {
...   |
298 | |             ZipImpl::try_fold(&mut self, init, NeverShortCircuit::wrap_mut_2(f)).0
    | |                                                                   |
    | |                                                                   |
    | |                                                                   function or associated item not found in `NeverShortCircuit<_>`
    | |                                                                   help: there is an associated function with a similar name: `wrap_mut_2_imp`
332 | |     };
333 | | }
333 | | }
    | |_- in this expansion of `zip_impl_general_defaults!`
...
377 |       zip_impl_general_defaults! {}
    |
   ::: library/core/src/ops/try_trait.rs:379:1
    |
    |
379 |   pub(crate) struct NeverShortCircuit<T>(pub T);
    |   -------------------------------------- function or associated item `wrap_mut_2` not found for this struct
For more information about this error, try `rustc --explain E0599`.
[RUSTC-TIMING] core test:false 7.338
error: could not compile `core` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3791353368/jobs/6446734888)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3791353368/jobs/6446734888","builder_name":"checks-actions"} -->

saethlin commented:
Whelp, that didn't work, but that also means it would have failed when approved anyway. @sarah-ek can you rebase this again and see if you can fix the compile error that appears?

JohnCSimon commented:
@sarah-ek 
ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

bors commented:
:umbrella: The latest upstream changes (presumably #107634) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
